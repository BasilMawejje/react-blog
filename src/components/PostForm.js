import React from 'react'
import { Field, reduxForm } from 'redux-form'

class PostForm extends React.Component {
    renderError = ({ error, touched }) => {
        if (error && touched) {
            return (
                <div className='ui error message'>
                    <div className='header'>{error}</div>
                </div>
            );
        }
    }

    renderInput = ({ input, label, meta }) => {
        return (
                <div className='field'>
                    <label>{label}</label>
                    <input {...input} autoComplete='off' />
                    {this.renderError(meta)}
                </div>
            );
    }

    onSubmit = (formValues) => {
        this.props.onSubmit(formValues);
    }

    render() {
        return (
            <form onSubmit={this.props.handleSubmit(this.onSubmit)} className='ui form error'>
                <Field name='title' component={this.renderInput} label='Enter title'/>
                <Field name='body' component={this.renderInput} label='Enter description' />
                <button className='ui button primary'>Submit</button>
            </form>
        );
    }
}

const validate = (formValues) => {
    let errors = {};

    if(!formValues.title) {
        errors.title = 'Please enter a title';
    }

    if(!formValues.body) {
        errors.body = 'Please enter a description';
    }
    
    return errors;
}

export default reduxForm({
    form: 'postForm',
    validate
})(PostForm);