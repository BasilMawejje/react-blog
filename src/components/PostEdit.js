import React from 'react'
import { connect } from 'react-redux'
import _ from 'lodash'

import { fetchPost, editPost } from '../actions'
import PostForm from './PostForm'

class PostEdit extends React.Component {
    componentDidMount() {
        this.props.fetchPost(this.props.match.params.id);
    }

    onSubmit = formValues => {
        this.props.editPost(this.props.match.params.id, formValues);
    };
    
    render() {
        if(!this.props.post) {
            return <div>Loading...</div>
        }

        return (
            <div>
                <h1>Edit Post</h1>
                <PostForm initialValues={_.pick(this.props.post, 'title', 'body')} onSubmit={this.onSubmit}/>
            </div>
        );
    };
}

const mapStateToProps = (state, ownProps) => {
    let id = ownProps.match.params.id;
    return { post: state.posts[id] }
}

export default connect(mapStateToProps, { fetchPost, editPost })(PostEdit);