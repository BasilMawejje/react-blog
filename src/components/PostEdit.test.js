import React from 'react';
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux'
import { shallowToJson } from 'enzyme-to-json';

import PostEdit from './PostEdit'
import PostForm from './PostForm'

configure({ adapter: new Adapter() });

const mockStore = configureStore();
const store = mockStore({posts: 
    [{ id: 1, title: 'Some post', body: 'lorem ipsum ...' }]
});

describe('PostEdit', () => {
    let wrapper; 

    beforeEach(() => {
        wrapper = shallow(
            <Provider store={store}>
                <PostEdit>
                    <PostForm />
                </PostEdit>
            </Provider>
        ).dive();
    });

    it('renders properly', () => { 
        expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
})