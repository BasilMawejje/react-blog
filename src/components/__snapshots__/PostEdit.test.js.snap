// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostEdit renders properly 1`] = `
<Connect(PostEdit)>
  <ReduxForm
    destroyOnUnmount={true}
    enableReinitialize={false}
    forceUnregisterOnUnmount={false}
    form="postForm"
    getFormState={[Function]}
    keepDirtyOnReinitialize={false}
    persistentSubmitErrors={false}
    pure={true}
    shouldAsyncValidate={[Function]}
    shouldError={[Function]}
    shouldValidate={[Function]}
    shouldWarn={[Function]}
    submitAsSideEffect={false}
    touchOnBlur={true}
    touchOnChange={false}
    updateUnregisteredFields={false}
    validate={[Function]}
  />
</Connect(PostEdit)>
`;
